Bio/Cluster/cluster.c

```diff
@@ -37,14 +37,29 @@
 /* ************************************************************************ */
 
 double
-mean(int n, double x[])
+mean(int n, double a[])
+// Add 4 elements at once instead of 1. The advantages are:
+// 1. less loop overhead
+// 2. compile with -O2 -> use SSE/AVX. 
+// 3. without AVX, still faster because 4 independent additions -> parallel instruction possible
+// 4. smaller floating-point error
 {
     double result = 0.;
     int i;
-
-    for (i = 0; i < n; i++) result += x[i];
-    result /= n;
-    return result;
+	double sum[4] = {0.,0.,0.,0.};
+	
+	int nstep4 = n - n % 4;
+	for (i=0; i<nstep4; i+=4) {	
+		sum[0] += a[i];
+		sum[1] += a[i + 1];
+		sum[2] += a[i + 2];
+		sum[3] += a[i + 3];
+	}
+
+    for (i = nstep4; i < n; i++) result += a[i];
+    result = result + sum[0] + sum[1] + sum[2] + sum[3];
```

Comment:
Should that be:

```c
result += sum[0] + sum[1] + sum[2] + sum[3];
```

Reply:
yes it's the same.  But the orignal code also wrote 
`result = result / n`
so I decided to keep the same style

Reply:
You seem to be talking about the next bit of code here? The original did ``result /= n;`` and ``return result`` as two lines - personally I think ``return result / n`` is you have it is just as clear and maybe faster.

Reply:
Oh okay. Do I need to change the code into "result += ..." and send a new PR, or could you change it for me? 
I'm new to github, so I'm sorry for asking :(

Reply:
If you want to make further changes, do another commit to the same branch, and push the updated branch to your github, and it will update this pull request. Very easy. But you might want to wait for Michiel's feedback first?

(And as long as you left the option ticked on the pull request, the Biopython maintainers can also update the branch on your GitHub biopython repository)

Reply:
On Mac at least it is indeed a lot faster, by a factor of about 2.5.
I agree with @peterjc  that `result += sum[0] + sum[1] + sum[2] + sum[3];` is better.
Is it better to define four variables sum1, sum2, sum3, sum4 instead of an array? Then we save an array lookup.

Reply:
**It might be worse,** depending on the compiler, but I guess a smart compiler will produce the same code.

By writing `sum[4]`, we declare to the compiler that we want 4 continuous addresses. In the main loop, this helps the compiler understand that "I want to sum 4 adjacents address with another 4 adjacent addresses", so it will use AVX-instruction immediately. In fact, this only need **1 array lookup instead of 4.**

If we write `sum1, ... sum4`, the (possibly stupid) compiler might not understand that we want it to vectorize this code.

Finally, 4 addresses next to each other in memory only require 1 main memory access, and 3 cache access (20x faster than RAM), instead of 4 RAM read. This is because x86 CPU always reads 64 bytes at once, so 4 values can be read with 1 array lookup.